service/core/v2ray/asset/asset.go

```diff
@@ -1,86 +1,121 @@
 package asset
 
 import (
+	"errors"
+	"github.com/adrg/xdg"
 	"github.com/muhammadmuzzammil1998/jsonc"
-	"github.com/v2rayA/v2rayA/common"
 	"github.com/v2rayA/v2rayA/common/files"
 	"github.com/v2rayA/v2rayA/conf"
 	"github.com/v2rayA/v2rayA/core/v2ray/where"
 	"github.com/v2rayA/v2rayA/pkg/util/log"
+	"io/fs"
 	"os"
 	"path"
 	"path/filepath"
+	"runtime"
 	"time"
 )
 
-func GetV2rayLocationAsset() (s string) {
-	var candidates = []string{
-		"/usr/local/share/v2ray",
-		"/usr/share/v2ray",
-		"/opt/share/v2ray",
-		"/usr/local/share/xray",
-		"/usr/share/xray",
-		"/opt/share/xray",
+func GetV2rayLocationAssetOverride() string {
+	return filepath.Join(xdg.RuntimeDir, "v2raya")
+}
+
+func GetV2rayLocationAsset(filename string) (string, error) {
+	variant, _, err := where.GetV2rayServiceVersion()
+	if err != nil {
+		variant = where.Unknown
+	}
+	var location string
+	var folder string
+	switch variant {
+	case where.V2ray:
+		location = "V2RAY_LOCATION_ASSET"
+		folder = "v2ray"
+	case where.Xray:
+		location = "XRAY_LOCATION_ASSET"
+		folder = "xray"
+	default:
+		location = "V2RAY_LOCATION_ASSET"
+		folder = "v2ray"
+	}
+	location = os.Getenv(location)
+	searchPaths := make([]string, 0)
+	if location != "" {
+		searchPaths = append(
+			searchPaths,
+			filepath.Join(location, filename),
+		)
 	}
-	var is bool
-	if ver, err := where.GetV2rayServiceVersion(); err == nil {
-		if is, err = common.VersionGreaterEqual(ver, "4.27.1"); is {
-			for _, c := range candidates {
-				if _, err := os.Stat(c); os.IsNotExist(err) {
-					continue
-				}
-				if _, err := os.Stat(path.Join(c, "geoip.dat")); os.IsNotExist(err) {
-					continue
-				}
-				s = c
-				break
+	if runtime.GOOS != "windows" {
+		searchPaths = append(
+			searchPaths,
+			filepath.Join("/usr/local/share", folder, filename),
+			filepath.Join("/usr/share", folder, filename),
+		)
+	}
+	if location != "" {
+		for _, searchPath := range searchPaths {
+			if _, err = os.Stat(searchPath); err != nil && errors.Is(err, fs.ErrNotExist) {
+				continue
 			}
+			return searchPath, nil
 		}
+		return searchPaths[0], nil
+	} else {
+		relpath := filepath.Join(folder, filename)
+		fullpath, err := xdg.SearchDataFile(relpath)
+		if err != nil {
+			fullpath, err = xdg.DataFile(relpath)
+			if err != nil {
+				return "", err
+			}
+		}
+		runtimepath, err := xdg.RuntimeFile(filepath.Join("v2raya", filename))
+		if err != nil {
+			return "", err
+		}
+		os.Remove(runtimepath)
+		err = os.Symlink(fullpath, runtimepath)
+		if err != nil {
+			return "", err
+		}
+		return fullpath, err
 	}
-	if s == "" {
-		// set as v2rayA config directory
-		s = conf.GetEnvironmentConfig().Config
-	}
-	return
 }
 
-func IsGFWListExists() bool {
-	_, err := os.Stat(path.Join(GetV2rayLocationAsset(), "LoyalsoldierSite.dat"))
+func IsV2rayAssetExists(filename string) bool {
```

Comment:
Sorry for my grammatical errors. Please help me rename it as DoesV2rayAssetExist, and the same below.